AntD

rymaju opened the pull request:
## Why

[Monday.com Ticket](https://code4community-team.monday.com/boards/XXXXXXXX/pulses/XXXXXXX)

What benefit does this bring to the end user? Or, what benefit does this bring to developers working in the codebase?

## This PR

Describe the changes required and any implementation choices you made to give context to reviewers. 

## Screenshots

Provide screenshots of any new components, styling changes, or pages. 

## Verification Steps

What steps did you take to verify your changes work? These should be clear enough for someone to be able to clone the branch and follow the steps themselves.

rymaju committed:
init

rymaju committed:
fix location navbar bug

rymaju committed:
lint

jackblanc commented on craco.config.js:
```diff
@@ -0,0 +1,18 @@
+      },
+    },
+  ],
+};
```
⛏️ newline

rymaju replied:
f

jackblanc commented on craco.config.js:
```diff
@@ -0,0 +1,18 @@
+      options: {
+        lessLoaderOptions: {
+          lessOptions: {
+            //This is where we override LESS global variables
```
❓ Are these variables indexable, like, in IntelliJ?

rymaju replied:
Not sure what you mean? We can import/require and format into whatever we want though!

jackblanc commented on craco.config.js:
```diff
@@ -0,0 +1,18 @@
+const CracoLessPlugin = require('craco-less');
```
❓ Why this syntax?

rymaju replied:
Because its just a regular node module, I think this is fine.

jackblanc commented on package.json:
```diff
@@ -3,6 +3,7 @@
   "private": true,
   "dependencies": {
+    "@craco/craco": "^5.7.0",
     "@material-ui/core": "^4.11.0",
```
🔪 🔪 🔪

jackblanc commented on package.json:
```diff
@@ -23,10 +26,9 @@
-    "build": "react-scripts build",
-    "test": "react-scripts test",
-    "eject": "react-scripts eject",
+    "start": "craco start",
```
Ah I didn't realize this was how significant craco was... interesting

jackblanc commented on src/App.tsx:
```diff
@@ -2,35 +2,45 @@ import React from 'react';
-import NavBar from './components/NavBar';
-
-import { ThemeProvider } from '@material-ui/core';
+// Import antd stylesheets
```
suuuuuper nitpick comment ⛏️  lowercase `import` 😂

jackblanc commented on src/App.tsx:
```diff
@@ -2,35 +2,45 @@ import React from 'react';
-
-import { ThemeProvider } from '@material-ui/core';
+// Import antd stylesheets
+import 'antd/dist/antd.css';
```
💭 Okay, so directly loading their distributed styles. Does this only have to be imported in App?

jackblanc commented on src/App.tsx:
```diff
@@ -2,35 +2,45 @@ import React from 'react';
+import 'antd/dist/antd.css';
+import './App.less';
+import Home from './containers/home/Home';
+import BlockTemplate from './containers/template-1-col-block/Template';
```
These are just example containers, yes?

jackblanc commented on src/App.tsx:
```diff
@@ -2,35 +2,45 @@ import React from 'react';
+import { Layout } from 'antd';
+const { Content } = Layout;
 
 const App: React.FC = () => {
   return (
```
```suggestion
const App: React.FC = () => (
```

jackblanc commented on src/components/Footer.tsx:
```diff
@@ -0,0 +1,11 @@
+const { Header, Footer, Sider, Content } = Layout;
+
+const NavBar: React.FC = () => {
+  const history = useHistory();
```
What's history used for here?

jackblanc commented on src/components/NavBar.tsx:
```diff
@@ -1,15 +1,59 @@
+const { Header, Footer, Sider, Content } = Layout;
 
 const NavBar: React.FC = () => {
+  const history = useHistory();
```
Do these have types?

rymaju replied:
![image](https://user-images.githubusercontent.com/23691775/94755221-c367c180-0361-11eb-89b8-c4de1fb22244.png)
Thats a no sadly.

jackblanc commented on src/components/NavBar.tsx:
```diff
@@ -1,15 +1,59 @@
+  const location = useLocation();
+  const path: string = location.pathname;
+
+  function pathnameToKey(pathname: string) {
```
💭 This could also be a StringMap (I think that's a standard type?)

rymaju replied:
oo Good idea!

jackblanc commented on src/components/NavBar.tsx:
```diff
@@ -1,15 +1,59 @@
+      >
+        <Menu.Item
+          key="1"
+          onClick={() => {
```
Hmm, I think there's a linting rule in TS against inline lambdas for performance reasons, so we may want to consider another approach/enable that. 
https://jonhilton.net/typescript-and-react-forbidden-lambdas/

rymaju replied:
Hmm ok, I think its less expressive though.

jackblanc commented on src/containers/template-24-col-grid/Template.tsx:
```diff
@@ -0,0 +1,111 @@
+            </Paragraph>
+          </Col>
+        </Row>
+        <Row gutter={16}>
+          <Col span={8}>
+            <Title level={3}>Grid allows </Title>
```
❓ what exactly do each of these do?

rymaju replied:
I think I covered this in the meeting

jackblanc commented on src/containers/template-24-col-grid/template.less:
```diff
@@ -0,0 +1,14 @@
+    border: 1px @primary-color solid;
+    height: 50px;
+    text-align: center;
+}
```
newline

jackblanc replied:
Hmmm, shouldn't lint enforce this? Might have to add `.less` to file extensions in lint command?

jackblanc reviewed (commented):
Awesome start on this! Left a bunch of questions. I also requested @bensenescu take a look as he's been working with Ant.D full-time.

rymaju committed:
add IBM plex sans

rymaju committed:
stretch content to fill screen

rymaju committed:
linting fixes

rymaju committed:
removed linting warnings

rymaju committed:
commented out test

rymaju committed:
test is back

rymaju committed:
fix prettier

rymaju commented:
- Added packages:
    - craco, craco-less
    - antd
- Removed MUI 
- Added basic AntD pages
- Moved files to the component/container structure
- fixed prettier commands

rymaju committed:
removed MUI

rymaju committed:
updated dependencies

jackblanc commented on package.json:
```diff
@@ -13,7 +13,9 @@
     "@types/react-router-dom": "^5.1.5",
+    "antd": "^4.6.6",
     "axios": "^0.19.2",
+    "craco-less": "^1.17.0",
```
@bensenescu you mentioned something about a `craco-antd` package, right?

jackblanc commented on src/App.less:
```diff
@@ -0,0 +1,4 @@
+.app-flex-container {
+  // need to use !important to overwrite .ant-layout
+  min-height: 100vh !important;
```
How often will we have to use `!important`? I've been told to always avoid it

rymaju replied:
Funny story...
Behind the scenes, AntD is loading its Layout css class. Due to the order in which the css is processed, our less files are not taking priority over the AntD default .ant-layout class.

Now there is a better way to do this, which is to implement the .ant-layout class ourselves as shown in documentation. I feel like this is just worse. Its literally copy pasting code to overwrite a single attribute (min-height). So I used !important instead.

I was afraid this might happen, and I wonder how this will affect our other "custom" css components. I imagine it would only be more difficult to customize other AntD components (without cascading global changes to all other AntD components).

jackblanc reviewed (approved):
There are still some open questions I have but this should get merged to unblock all other development
